gnmi/gnmit/gnmit.go

```diff
@@ -0,0 +1,338 @@
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to listen: %v", err)
+	}
+
+	go srv.Serve(lis)
+	c.stopFn = srv.GracefulStop
+	return c, lis.Addr().String(), nil
+}
+
+type populateDefaultser interface {
+	PopulateDefaults()
+}
+
+// New returns a new collector that listens on the specified addr (in the form host:port),
+// supporting a single downstream target named hostname. sendMeta controls whether the
+// metadata *other* than meta/sync and meta/connected is sent by the collector.
+//
+// New returns the new collector, the address it is listening on in the form hostname:port
+// or any errors encounted whilst setting it up.
+//
+// NewSettable is different from New in that the returned collector is
+// schema-aware and supports gNMI Set. Currently it is not possible to change
+// the schema of a Collector after it is created.
+func NewSettable(ctx context.Context, addr string, hostname string, sendMeta bool, schema *ytypes.Schema, tasks []Task, opts ...grpc.ServerOption) (*Collector, string, error) {
+	if !schema.IsValid() {
+		return nil, "", fmt.Errorf("cannot obtain valid schema for GoStructs: %v", schema)
+	}
+	// Initialize the root with default values.
+	schema.Root.(populateDefaultser).PopulateDefaults()
+	if err := schema.Root.Validate(); err != nil {
+		return nil, "", fmt.Errorf("default root of input schema fails validation: %v", err)
+	}
+
+	// FIXME(wenbli): initialize the collector with default values.
+	collector, addr, err := New(ctx, addr, hostname, sendMeta, tasks, opts...)
+	if err != nil {
+		return nil, "", err
+	}
+	collector.schema = schema
+	return collector, addr, nil
+}
+
+// Stop halts the running collector.
+func (c *Collector) Stop() {
+	c.stopFn()
+}
+
+// handleUpdate handles an input gNMI SubscribeResponse that is received by
+// the target.
+func (c *Collector) handleUpdate(resp *gpb.SubscribeResponse) error {
+	t := c.cache.GetTarget(c.name)
+	switch v := resp.Response.(type) {
+	case *gpb.SubscribeResponse_Update:
+		t.GnmiUpdate(v.Update)
+	case *gpb.SubscribeResponse_SyncResponse:
+		t.Sync()
+	case *gpb.SubscribeResponse_Error:
+		return fmt.Errorf("error in response: %s", v)
+	default:
+		return fmt.Errorf("unknown response %T: %s", v, v)
+	}
+	return nil
+}
+
+// Collector is a basic gNMI target that supports only the Subscribe
+// RPC, and acts as a cache for exactly one target.
+type Collector struct {
+	cache  *cache.Cache
+	smu    sync.Mutex
+	schema *ytypes.Schema
+	// name is the hostname of the client.
+	name string
+	// inCh is a channel use to write new SubscribeResponses to the client.
+	inCh chan *gpb.SubscribeResponse
+	// stopFn is the function used to stop the server.
+	stopFn func()
+}
+
+// TargetUpdate provides an input gNMI SubscribeResponse to update the
+// cache and clients with.
+func (c *Collector) TargetUpdate(m *gpb.SubscribeResponse) {
+	c.inCh <- m
+}
+
+func (s *GNMIServer) Set(ctx context.Context, req *gpb.SetRequest) (*gpb.SetResponse, error) {
```

Comment:
Should we break this out into a function that we can pull out and test separately? It looks like we're getting to a bigger function here.

I suggest that since we're in prototype mode, we file issues that say "this thing should probably be fixed like this", and then we go back and fix those once we're more confident with the design.

Reply:
This is being tested by `TestONCESet`. I added TODO to handle refactoring this.